File: web/src/main/java/net/danil/web/problem/service/TestRunnerChannelService.java
package net.danil.web.problem.service;

import lombok.RequiredArgsConstructor;
import net.danil.web.problem.dto.TestResult;
import net.danil.web.problem.model.SolutionResult;
import net.danil.web.problem.repository.SolutionRepository;
import net.danil.web.problem.repository.SolutionResultRepository;
import net.danil.web.statistics.StatisticsRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.kafka.support.KafkaHeaders;
import org.springframework.messaging.Message;
import org.springframework.messaging.handler.annotation.Header;
import org.springframework.messaging.handler.annotation.Payload;
import org.springframework.messaging.support.GenericMessage;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;
import reactor.core.publisher.MonoSink;

import java.util.concurrent.ConcurrentHashMap;

@Service
@RequiredArgsConstructor
public class TestRunnerChannelService {
    public static final String TOPIC_NAME = "test-result-topic";

    private final ConcurrentHashMap<String, MonoSink<Message<?>>> messageHandlers = new ConcurrentHashMap<>();
    private final Logger logger = LoggerFactory.getLogger(TestRunnerChannelService.class);
    private final SolutionResultRepository solutionResultRepository;
    private final SolutionRepository solutionRepository;
    private final TestResultAnalyzerService testResultAnalyzerService;
    private final StatisticsRepository statisticsRepository;

    @KafkaListener(topics = TOPIC_NAME)
    protected void listen(@Payload TestResult testResult, @Header(value = KafkaHeaders.RECEIVED_KEY, required = false) String messageId) {
        if (messageId == null) {
            logger.error("received null message id");
            return;
        }
        logger.info("received result forId({}): {}", messageId, testResult);
        final var sink = messageHandlers.get(messageId);
        try {
            final var verdict = testResultAnalyzerService.judgeResults(testResult);
            final var solution = solutionRepository.findById(testResult.solutionId()).get();

            SolutionResult solutionResult = new SolutionResult(solution.getId(), testResult.tests(), testResult.failures(), testResult.errors(), testResult.statusCode(), testResult.time(),
                    testResult.logs(), testResult.xml(), verdict.solved(), solution);
            solutionResultRepository.save(solutionResult);
            solution.setSolutionResult(solutionResult);
            statisticsRepository.updateUserStatistics(solution.getUser().getId());

            sink.success(new GenericMessage<>(solution));
        } catch (Exception e) {
            sink.error(new RuntimeException(testResult.toString(), e));
        }
    }

    public Mono<Message<?>> subscribe(String id) {
        return Mono.create(sink -> messageHandlers.put(id, sink));
    }
}

File: web/src/main/java/org/supremecode/web/service/TestRunnerChannelService.kt
package org.supremecode.web.service

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.kafka.support.KafkaHeaders
import org.springframework.messaging.handler.annotation.Header
import org.springframework.messaging.handler.annotation.Payload
import org.springframework.transaction.annotation.Transactional
import org.springframework.transaction.support.TransactionSynchronization
import org.springframework.transaction.support.TransactionSynchronizationManager
import org.supremecode.shared.TestResultMessage
import org.supremecode.shared.mdcContext
import org.supremecode.web.domain.SolutionResult
import org.supremecode.web.repository.SolutionRepository
import org.supremecode.web.repository.SolutionResultRepository

open class TestRunnerChannelService(
    val solutionRepository: SolutionRepository,
    val solutionResultRepository: SolutionResultRepository,
    val businessMetrics: BusinessMetrics,
    val solutionResultNotificationService: SolutionResultNotificationService,
) {
    val logger: Logger = LoggerFactory.getLogger(TestRunnerChannelService::class.java)

    @KafkaListener(topics = [TOPIC_NAME])
    @Transactional
    open fun listen(
        @Payload testResult: TestResultMessage,
        @Header(value = KafkaHeaders.RECEIVED_KEY, required = true) messageId: String
    ) {
        mdcContext(
            "user_id" to testResult.userId,
            "problem_id" to testResult.problemId,
            "solution_id" to testResult.solutionId,
            "language_id" to testResult.languageId,
            "kafka_message_id" to messageId,
            "status_code" to testResult.statusCode,
            "solved" to testResult.solved,
        ).use {
            logger.info("test result received")
            val solution = solutionRepository.findById(testResult.solutionId).get()
            val solutionResult = SolutionResult(
                solution, testResult.statusCode, testResult.solved, testResult.total,
                testResult.failures, testResult.errors, testResult.testCases ?: emptyList()
            )
            solutionResultRepository.save(solutionResult)

            val language = solution.problemLanguage.languageId
            val latencyMillis = solutionResult.createdAt.time - solution.createdAt.time
            businessMetrics.recordResult(language, testResult.solved, testResult.statusCode)
            businessMetrics.recordResultLatency(language, testResult.solved, testResult.statusCode, latencyMillis)
            mdcContext("duration_ms" to latencyMillis).use { logger.info("test result persisted") }
            publishAfterCommit(testResult)
        }
    }

    private fun publishAfterCommit(testResult: TestResultMessage) {
        if (!TransactionSynchronizationManager.isSynchronizationActive()) {
            solutionResultNotificationService.publish(testResult)
            return
        }

        TransactionSynchronizationManager.registerSynchronization(object : TransactionSynchronization {
            override fun afterCommit() {
                solutionResultNotificationService.publish(testResult)
            }
        })
    }

    companion object {
        const val TOPIC_NAME: String = "test-result-topic"
    }
}
